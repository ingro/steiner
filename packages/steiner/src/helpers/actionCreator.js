// import snakeCase from 'lodash/snakeCase';
import upperFirst from 'lodash/upperFirst';
import defaults from 'lodash/defaults';
import trim from 'lodash/trim';

function addAsyncGroup(resource, actionTypes, group, config) {
    // const upperResource = snakeCase(resource).toUpperCase();
    const upperGroup    = group.toUpperCase();

    resource = trim(resource);

    // const start        = `${upperResource}_${upperGroup}`;
    // const success      = `${upperResource}_${upperGroup}_SUCCESS`;
    // const fail         = `${upperResource}_${upperGroup}_FAIL`;
    const start        = `${resource}/${upperGroup}`;
    const success      = `${resource}/${upperGroup}_SUCCESS`;
    const fail         = `${resource}/${upperGroup}_FAIL`;
    const startAlias   = group;
    const successAlias = `${group}Success`;
    const failAlias    = `${group}Fail`;

    actionTypes[start]   = start;
    actionTypes[success] = success;
    actionTypes[fail]    = fail;

    if (config.addAlias) {
        actionTypes[startAlias]   = start;
        actionTypes[successAlias] = success;
        actionTypes[failAlias]    = fail;
    }
}

export function createActionTypes(resource, config) {
    if (resource == null) {
        throw new Error('Expected resource');
    }

    config = config || {};

    if (config.addAlias == null) {
        config.addAlias = true;
    }

    resource = trim(resource);

    if (resource === '') {
        throw new Error('Expected resource');
    }

    const actionTypes = {};

    addAsyncGroup(resource, actionTypes, 'list', config);
    addAsyncGroup(resource, actionTypes, 'fetch', config);
    addAsyncGroup(resource, actionTypes, 'create', config);
    addAsyncGroup(resource, actionTypes, 'update', config);
    addAsyncGroup(resource, actionTypes, 'delete', config);

    // const upperResource = snakeCase(resource).toUpperCase();

    const resetCurrent = `${resource}/RESET_CURRENT`;
    const changePage   = `${resource}/CHANGE_PAGE`;
    const updateFilter = `${resource}/UDATE_FILTER`;
    const changeOrder  = `${resource}/CHANGE_ORDER`;
    // const inputQuery   = `${upperResource}_INPUT_QUERY`;

    actionTypes.resetCurrent = resetCurrent;
    actionTypes[resetCurrent] = resetCurrent;

    actionTypes.changePage = changePage;
    actionTypes[changePage] = changePage;

    actionTypes.updateFilter = updateFilter;
    actionTypes[updateFilter] = updateFilter;

    actionTypes.changeOrder = changeOrder;
    actionTypes[changeOrder] = changeOrder;

    // actionTypes.inputQuery = inputQuery;
    // actionTypes[inputQuery] = inputQuery;

    return actionTypes;
}

function generateDefaultMessages(resource) {
    return {
        createSuccess: `${resource} created with success!`,
        createFail: `An error occured while creating ${resource}`,
        updateSuccess: `${resource} updated successfully!`,
        updateFail: `An error occured while updating ${resource}`,
        deleteSuccess: `${resource} deleted with success!`,
        deleteFail: `An error occured while deleting ${resource}`
    };
};

function generateSuccessNotification(message, title = 'Hooray!') {
    return {
        title,
        message,
        status: 'success'
    }
}

function generateFailNotification(message, title = 'Oh snap!') {
    return {
        title,
        message,
        status: 'error'
    }
}

export function createActions(resource, actionTypes, messages = {}) {
    if (resource == null) {
        throw new Error('Expected resource');
    }

    resource = trim(upperFirst(resource));

    const defaultMessages = generateDefaultMessages(resource);

    defaults(messages, defaultMessages);

    const actions = {};

    actions[`fetch`] = function(id) {
        return {
            type: actionTypes.fetch,
            payload: {
                id
            }
        };
    }

    actions[`list`] = function(filters) {
        return {
            type: actionTypes.list,
            loadingBar: 'show'
        };
    }

    actions[`create`] = function(data) {
        return {
            type: actionTypes.create,
            payload: {
                data
            }
        }
    }

    actions['createSuccess'] = function(response) {
        return {
            type: actionTypes.createSuccess,
            payload: response,
            notification: generateSuccessNotification(messages.createSuccess)
        }
    }

    actions['createFail'] = function(error) {
        return {
            type: actionTypes.createFail,
            error,
            notification: generateFailNotification(messages.createFail)
        }
    }

    actions[`update`] = function(id, data/*, resolve, reject*/) {
        return {
            type: actionTypes.update,
            payload: {
                id,
                data
                /*resolve,
                reject*/
            }
        }
    }

    actions['updateSuccess'] = function(response) {
        return {
            type: actionTypes.updateSuccess,
            payload: response,
            notification: generateSuccessNotification(messages.updateSuccess)
        }
    }

    actions['updateFail'] = function(error) {
        return {
            type: actionTypes.updateFail,
            error,
            notification: generateFailNotification(messages.updateFail)
        }
    }

    actions[`delete`] = function(id) {
        return {
            type: actionTypes.delete,
            payload: {
                id
            }
        }
    }

    actions['deleteSuccess'] = function(payload) {
        return {
            type: actionTypes.deleteSuccess,
            payload,
            notification: generateSuccessNotification(messages.deleteSuccess)
        }
    }

    actions['deleteFail'] = function(error) {
        return {
            type: actionTypes.deleteFail,
            error,
            notification: generateFailNotification(messages.deleteFail)
        }
    }

    actions[`resetCurrent`] = function() {
        return {
            type: actionTypes.resetCurrent
        };
    }

    // actions[`updateQuery`] = function(q) {
    //     return {
    //         type: actionTypes.inputQuery,
    //         payload: {
    //             q
    //         }
    //     };
    // }

    actions['changePage'] = function(page) {
        return {
            type: actionTypes.changePage,
            payload: {
                page
            }
        }
    }

    actions['changeOrder'] = function(key, direction) {
        return {
            type: actionTypes.changeOrder,
            payload: {
                key,
                direction
            }
        }
    }

    actions['updateFilter'] = function(key, value) {
        return {
            type: actionTypes.updateFilter,
            payload: {
                key,
                value
            }
        };
    }

    return actions;
}

export default {
    createActions,
    createActionTypes
};