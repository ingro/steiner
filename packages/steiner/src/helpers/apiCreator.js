import _ from 'lodash';

export const DEFAULT_PARAMS_MAP = {
    q: 'q',
    _page: 'page',
    _limit: 'perPage',
    _sort: 'orderKey',
    _order: 'orderDirection'
};

export function buildParams(filters, paramsMap = {}) {
    const params = {};

    _.defaults(paramsMap, DEFAULT_PARAMS_MAP);

    _.forOwn(paramsMap, (value, key) => {
        if (typeof value === 'function') {
            params[key] = value(filters);
        } else if (key !== '') {
            params[key] = _.get(filters, value);
        }
    });

    return params;
}

const defaultOptions = {
    omit: []
};

export default function createApi(endpoint, client, paramsMap = {}, options = {}) {
    _.defaults(options, defaultOptions);
    const routes = {};

    const [uri, qs] = endpoint.split('?');

    routes.fetch = function(id) {
        return client({
            url: `/${uri}/${id}${qs ? '?' + qs : ''}`
        });
    }

    routes.list = function(filters) {
        return client({
            url: `/${uri}${qs ? '?' + qs : ''}`,
            params: buildParams(filters, paramsMap)
        });
    }

    routes.create = function(data) {
        return client({
            url: `/${uri}`,
            method: 'post',
            data
        });
    }

    routes.update = function(id, data) {
        return client({
            url: `/${uri}/${id}`,
            method: 'patch',
            data
        });
    }

    routes.delete = function(id) {
        return client({
            url: `/${uri}/${id}`,
            method: 'delete'
        });
    }

    return _.omit(routes, options.omit);
}