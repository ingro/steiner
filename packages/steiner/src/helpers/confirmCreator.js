import { addNotification as notify } from 'reapop';
import _ from 'lodash';

import defaultMessages from '../messages/en';

const defaultConfirmOptions = {
    status: 'warning',
    position: 'tc'
};

export default function createConfirm(options = {}) {
    _.defaults(options, defaultConfirmOptions, defaultMessages.messages.confirmDialog);

    return notify({
        title: options.title,
        message: options.message,
        status: options.status,
        dismissible: false,
        dismissAfter: 0,
        position: options.position,
        buttons: [
            {
                name: options.btn_primary_text,
                primary: true,
                onClick: () => {
                    options.onSuccess();
                }
            },
            {
                name: options.btn_secondary_text,
                primary: false
            }
        ]
    });
}