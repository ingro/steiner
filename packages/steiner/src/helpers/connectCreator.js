import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

export function connectList(component, actions, selectors) {
    function mapStateToProps(state) {
        const list = selectors.listSelector(state);

        return {
            ...list,
            items: selectors.itemsSelector(state),
            selected: selectors.getSelectedId(state),
            component
        };
    }

    function mapDispatchToProps(dispatch) {
        const boundedActions = bindActionCreators(actions, dispatch);

        return {
            ...boundedActions,
            dispatch
        }
    }

    return connect(mapStateToProps, mapDispatchToProps)(component);
}

export function connectItem(component, actions, selectors) {
    function mapStateToProps(state, ownProps) {
        const current = selectors.currentSelector(state);
        const isNew = ownProps.params.id === 'create';

        return {
            ...current,
            item: isNew ? {} : current.item
        };
    }

    function mapDispatchToProps(dispatch) {
        return bindActionCreators(actions, dispatch);
    }

    return connect(mapStateToProps, mapDispatchToProps)(component);
}

export default {
    connectItem,
    connectList
};