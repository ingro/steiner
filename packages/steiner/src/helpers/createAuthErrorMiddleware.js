import { addNotification as notify } from 'reapop';

import { actions } from '../auth/actions';
import { getLanguage, getTranslations } from '../settings/reducer';

function createAuthErrorMiddleware(options = {}) {
    const authErrorMiddleware = store => next => action => {
        if (action.error && action.error.response && action.error.response.status === 401) {
            let title;
            let message;

            const state = store.getState();

            if (options.messages) {
                const language = getLanguage(state);

                title = options.messages[language].title;
                message = options.messages[language].message;
            } else {
                const translations = getTranslations(state);

                title = translations.messages.notifications.titles.fail;
                message = translations.messages.notifications.authError;
            }

            store.dispatch(notify({
                title,
                message,
                status: 'error',
                position: 'tc',
                dismissAfter: 0
            }));

            store.dispatch(actions.logoutRequestSuccess());
        }

        return next(action);
    };

    return authErrorMiddleware;  
}

export default createAuthErrorMiddleware;
