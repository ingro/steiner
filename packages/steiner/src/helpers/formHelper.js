import { createFormAction } from './reduxFormSaga';

const identity = data => data;

export function createSubmit(actionTypes, dataTransformer = identity) {
    return function submit(data, dispatch) {
        const action = data.id
            ? createFormAction(actionTypes.update, [actionTypes.updateSuccess, actionTypes.updateFail])
            : createFormAction(actionTypes.create, [actionTypes.createSuccess, actionTypes.createFail]);

        return action(dataTransformer(data), dispatch);
    }
}

export default {
    createSubmit
};