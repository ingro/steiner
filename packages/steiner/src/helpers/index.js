import actionCreatorModule from './actionCreator';
import apiCreatorModule from './apiCreator';
import connectCreatorModule from './connectCreator';
import formHelperModule from './formHelper';
import notificationMiddlewareModule from './notificationMiddleware';
import reducerCreatorModule from './reducerCreator';
import reduxFormSagaModule from './reduxFormSaga';
import routeCreatorModule from './routeCreator';
import sagaCreatorModule from './sagaCreator';

export const actionCreator = actionCreatorModule;
export const apiCreator = apiCreatorModule;
export const connectCreator = connectCreatorModule;
export const formHelper = formHelperModule;
export const notificationMiddleware = notificationMiddlewareModule;
export const reducerCreator = reducerCreatorModule;
export const reduxFormSaga = reduxFormSagaModule;
export const routeCreator = routeCreatorModule;
export const sagaCreator = sagaCreatorModule;