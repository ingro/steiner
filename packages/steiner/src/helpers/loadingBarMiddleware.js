import { showLoading, hideLoading } from 'react-redux-loading-bar';

const loadingBarMiddleware = store => next => action => {
    if (action.loadingBar) {
        store.dispatch(action.loadingBar === 'show' ? showLoading() : hideLoading());
    }

    return next(action);
};

export default loadingBarMiddleware;