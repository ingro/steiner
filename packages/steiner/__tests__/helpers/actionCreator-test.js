import { createActions, createActionTypes } from '../../src/helpers/actionCreator';

describe('createActionTypes', () => {
    it('generates the correct action types', () => {
        const actionTypes = createActionTypes('posts');

        expect(actionTypes).toMatchSnapshot();
    });
});

describe('createActions', () => {
    it('generates the correct actions', () => {
        const actionTypes = createActionTypes('posts');

        const actions = createActions('posts', actionTypes);

        expect(actions.fetch(42)).toMatchSnapshot();
        expect(actions.list()).toMatchSnapshot();
        expect(actions.create({ foo: 'bar' })).toMatchSnapshot();
        expect(actions.createSuccess({ foo: 'bar' }, { type: 'success' })).toMatchSnapshot();
        expect(actions.createFail({ foo: 'bar' }, { type: 'fail' })).toMatchSnapshot();
        expect(actions.update(42, { foo: 'bar' })).toMatchSnapshot();
        expect(actions.updateSuccess({ foo: 'bar' }, { type: 'success' })).toMatchSnapshot();
        expect(actions.updateFail({ foo: 'bar' }, { type: 'fail' })).toMatchSnapshot();
        expect(actions.delete(42)).toMatchSnapshot();
        expect(actions.deleteSuccess({ foo: 'bar' }, { type: 'success' })).toMatchSnapshot();
        expect(actions.deleteFail({ foo: 'bar' }, { type: 'fail' })).toMatchSnapshot();
        expect(actions.resetCurrent()).toMatchSnapshot();
        expect(actions.changePage(3)).toMatchSnapshot();
        expect(actions.changeOrder('name', 'DESC')).toMatchSnapshot();
        expect(actions.updateFilter('q', 'foo')).toMatchSnapshot();
        expect(actions.syncFilters('q', 'bar')).toMatchSnapshot();
        expect(actions.checkFilterSync('q', 'baz')).toMatchSnapshot();
        expect(actions.select(42)).toMatchSnapshot();
        expect(actions.deselect(42)).toMatchSnapshot();
        expect(actions.selectAll()).toMatchSnapshot();
        expect(actions.deselectAll()).toMatchSnapshot();
        expect(actions.resetFilters()).toMatchSnapshot();
        expect(actions.setFilters({ foo: 'bar' })).toMatchSnapshot();
    });
});