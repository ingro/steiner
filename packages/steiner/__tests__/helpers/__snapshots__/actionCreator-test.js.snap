exports[`createActionTypes generates the correct action types 1`] = `
Object {
  "changeOrder": "posts/CHANGE_ORDER",
  "changePage": "posts/CHANGE_PAGE",
  "checkFilterSync": "posts/CHECK_FILTER_SYNC",
  "create": "posts/CREATE",
  "createFail": "posts/CREATE_FAIL",
  "createSuccess": "posts/CREATE_SUCCESS",
  "delete": "posts/DELETE",
  "deleteFail": "posts/DELETE_FAIL",
  "deleteSuccess": "posts/DELETE_SUCCESS",
  "deselect": "posts/DESELECT",
  "deselectAll": "posts/DESELECT_ALL",
  "fetch": "posts/FETCH",
  "fetchFail": "posts/FETCH_FAIL",
  "fetchSuccess": "posts/FETCH_SUCCESS",
  "filterCollection": "posts/FILTER_COLLECTION",
  "list": "posts/LIST",
  "listFail": "posts/LIST_FAIL",
  "listSuccess": "posts/LIST_SUCCESS",
  "posts/CHANGE_ORDER": "posts/CHANGE_ORDER",
  "posts/CHANGE_PAGE": "posts/CHANGE_PAGE",
  "posts/CHECK_FILTER_SYNC": "posts/CHECK_FILTER_SYNC",
  "posts/CREATE": "posts/CREATE",
  "posts/CREATE_FAIL": "posts/CREATE_FAIL",
  "posts/CREATE_SUCCESS": "posts/CREATE_SUCCESS",
  "posts/DELETE": "posts/DELETE",
  "posts/DELETE_FAIL": "posts/DELETE_FAIL",
  "posts/DELETE_SUCCESS": "posts/DELETE_SUCCESS",
  "posts/DESELECT": "posts/DESELECT",
  "posts/DESELECT_ALL": "posts/DESELECT_ALL",
  "posts/FETCH": "posts/FETCH",
  "posts/FETCH_FAIL": "posts/FETCH_FAIL",
  "posts/FETCH_SUCCESS": "posts/FETCH_SUCCESS",
  "posts/FILTER_COLLECTION": "posts/FILTER_COLLECTION",
  "posts/LIST": "posts/LIST",
  "posts/LIST_FAIL": "posts/LIST_FAIL",
  "posts/LIST_SUCCESS": "posts/LIST_SUCCESS",
  "posts/RESET_CURRENT": "posts/RESET_CURRENT",
  "posts/RESET_FILTERS": "posts/RESET_FILTERS",
  "posts/SELECT": "posts/SELECT",
  "posts/SELECT_ALL": "posts/SELECT_ALL",
  "posts/SET_FILTERS": "posts/SET_FILTERS",
  "posts/SYNC_FILTERS": "posts/SYNC_FILTERS",
  "posts/UDATE_FILTER": "posts/UDATE_FILTER",
  "posts/UPDATE": "posts/UPDATE",
  "posts/UPDATE_FAIL": "posts/UPDATE_FAIL",
  "posts/UPDATE_SUCCESS": "posts/UPDATE_SUCCESS",
  "resetCurrent": "posts/RESET_CURRENT",
  "resetFilters": "posts/RESET_FILTERS",
  "select": "posts/SELECT",
  "selectAll": "posts/SELECT_ALL",
  "setFilters": "posts/SET_FILTERS",
  "syncFilters": "posts/SYNC_FILTERS",
  "update": "posts/UPDATE",
  "updateFail": "posts/UPDATE_FAIL",
  "updateFilter": "posts/UDATE_FILTER",
  "updateSuccess": "posts/UPDATE_SUCCESS",
}
`;

exports[`createActions generates the correct actions 1`] = `
Object {
  "payload": Object {
    "id": 42,
  },
  "type": "posts/FETCH",
}
`;

exports[`createActions generates the correct actions 2`] = `
Object {
  "loadingBar": "show",
  "type": "posts/LIST",
}
`;

exports[`createActions generates the correct actions 3`] = `
Object {
  "payload": Object {
    "data": Object {
      "foo": "bar",
    },
  },
  "type": "posts/CREATE",
}
`;

exports[`createActions generates the correct actions 4`] = `
Object {
  "notification": Object {
    "type": "success",
  },
  "payload": Object {
    "foo": "bar",
  },
  "type": "posts/CREATE_SUCCESS",
}
`;

exports[`createActions generates the correct actions 5`] = `
Object {
  "error": Object {
    "foo": "bar",
  },
  "notification": Object {
    "type": "fail",
  },
  "type": "posts/CREATE_FAIL",
}
`;

exports[`createActions generates the correct actions 6`] = `
Object {
  "payload": Object {
    "data": Object {
      "foo": "bar",
    },
    "id": 42,
  },
  "type": "posts/UPDATE",
}
`;

exports[`createActions generates the correct actions 7`] = `
Object {
  "notification": Object {
    "type": "success",
  },
  "payload": Object {
    "foo": "bar",
  },
  "type": "posts/UPDATE_SUCCESS",
}
`;

exports[`createActions generates the correct actions 8`] = `
Object {
  "error": Object {
    "foo": "bar",
  },
  "notification": Object {
    "type": "fail",
  },
  "type": "posts/UPDATE_FAIL",
}
`;

exports[`createActions generates the correct actions 9`] = `
Object {
  "payload": Object {
    "id": 42,
  },
  "type": "posts/DELETE",
}
`;

exports[`createActions generates the correct actions 10`] = `
Object {
  "notification": Object {
    "type": "success",
  },
  "payload": Object {
    "foo": "bar",
  },
  "type": "posts/DELETE_SUCCESS",
}
`;

exports[`createActions generates the correct actions 11`] = `
Object {
  "error": Object {
    "foo": "bar",
  },
  "notification": Object {
    "type": "fail",
  },
  "type": "posts/DELETE_FAIL",
}
`;

exports[`createActions generates the correct actions 12`] = `
Object {
  "type": "posts/RESET_CURRENT",
}
`;

exports[`createActions generates the correct actions 13`] = `
Object {
  "payload": Object {
    "page": 3,
  },
  "type": "posts/CHANGE_PAGE",
}
`;

exports[`createActions generates the correct actions 14`] = `
Object {
  "payload": Object {
    "direction": "DESC",
    "key": "name",
  },
  "type": "posts/CHANGE_ORDER",
}
`;

exports[`createActions generates the correct actions 15`] = `
Object {
  "payload": Object {
    "key": "q",
    "value": "foo",
  },
  "type": "posts/UDATE_FILTER",
}
`;

exports[`createActions generates the correct actions 16`] = `
Object {
  "payload": "q",
  "type": "posts/SYNC_FILTERS",
}
`;

exports[`createActions generates the correct actions 17`] = `
Object {
  "payload": "q",
  "type": "posts/CHECK_FILTER_SYNC",
}
`;

exports[`createActions generates the correct actions 18`] = `
Object {
  "payload": Array [
    42,
  ],
  "type": "posts/SELECT",
}
`;

exports[`createActions generates the correct actions 19`] = `
Object {
  "payload": Array [
    42,
  ],
  "type": "posts/DESELECT",
}
`;

exports[`createActions generates the correct actions 20`] = `
Object {
  "type": "posts/SELECT_ALL",
}
`;

exports[`createActions generates the correct actions 21`] = `
Object {
  "type": "posts/DESELECT_ALL",
}
`;

exports[`createActions generates the correct actions 22`] = `
Object {
  "type": "posts/RESET_FILTERS",
}
`;

exports[`createActions generates the correct actions 23`] = `
Object {
  "payload": Object {
    "foo": "bar",
  },
  "type": "posts/SET_FILTERS",
}
`;
